staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go

```diff
@@ -137,10 +137,19 @@ func UnsafeGuessKindToResource(kind schema.GroupVersionKind) ( /*plural*/ schema
 		}
 	}
 
-	switch string(singularName[len(singularName)-1]) {
-	case "s":
+	switch {
```

Comment:
/hold

This isn't a compatible change to make to this function. Anything using this function to auto-generate plural names from singular names would break when re-generating an existing object that this change impacts.

I think we should remove all non-test in-tree calls to this function, mark it deprecated, file issues to public callers we find, and consider removing it in a future release.

If we want a *new* function that takes a singular name and pluralizes it that has this expanded behavior from the start (with a signature like `AttemptToPluralize(singular string) (plural string)`), I guess I wouldn't object, but that doesn't need to live in k8s.io/apimachinery

Reply:
I believe that the use that bit me _was_ in test coffee without an apiserver loaded. I can try to dig out the code in question; I believe it tries to pluralize "Gateway" as "Gatewaies", which was very confusing when I finally debugged it.

Removing this function would also be acceptable to me, but it's currently a trap for anyone testing the Gateway API.

Reply:
@evankanderson Can you dig that test code up? I'm positive I've also encountered this issue recently but can't find the relevant test code that would manifest the problem itself (other than calling `UnsafeGuessKindToResource` with Gateway type).

Reply:
I believe that we were getting hung up in `tracker.Add` from this code:

https://github.com/kubernetes/client-go/blob/master/testing/fixture.go#L341

I also see some questionable usage in https://github.com/knative/pkg/blob/main/resolver/addressable_resolver.go#L167 and https://github.com/knative/pkg/blob/main/kref/knative_reference.go#L53, but the usage of `ObjectTracker` from `client-go/testing` seems most likely to cause problems, because I'm guessing that package is used many places.

You can see my work-around here:

https://github.com/knative-sandbox/net-gateway-api/blob/main/pkg/reconciler/ingress/ingress_test.go#L404